### PR TITLE
NixOS container switched to Ubuntu container with Nix installed

### PR DESCRIPTION
**Pull request by rickvanprim**
Apologies for yet another PR.  Is there a better way to test this?

### CODE REVIEW

**Comment by meteorcloudy** [comment]
How did you find out those issues? Can you build the image locally and use that to test the build you want to run first?

**Comment by rickvanprim** [comment]
I've been finding those issues via the BuildKite runs on https://github.com/bazelbuild/rules_rust/pull/1979.  For this latest PR, I've built and run the container, and in the container, I've cloned and built the example from the `rules_rust` PR.  It's not a perfect recreation of the BuildKite execution, but the Python `bazelci.py` seemed to execute fine in the last attempt, and I've gotten the example building (with some "fun" Nix realizations along the way 😅).

I switched from `bazelisk` to `bazel_6` (which is currently Bazel 6.4.0), as that's considerably less against the grain (the former is trying to run a prebuilt non-Nix Bazel on Nix, which is using `nix-ld` to try to reactively fix that).  @UebelAndre what are your thoughts?  I can presumably get the `bazelisk` approach to work, but it might be brittle.  The last issue with it seems to be some sort of `glibc` mismatch, so potentially solvable with a bit more playing with it.

If we keep the switch to `bazel`, I'm thinking I should remove the `nix-ld` stuff, as I wouldn't expect this Nix example (or any other Nix examples) to be using prebuilt binaries that aren't supplied through Nix.

**Comment by meteorcloudy** [comment]
@rickvanprim Thanks for testing it! Let me know when you think it is ready.

**Comment by UebelAndre** [comment]
I think it's pretty important that the package has bazelisk installed so the image doesn't need to be updated when bazel releases happen. But installing a bazel package for accurate system dependencies sounds fine to me!

**Comment by rickvanprim** [comment]
@UebelAndre unfortunately Nix doesn't really work like that.  Nix doesn't have the concept of system wide dependencies.  Instead the Bazel package in Nix is compiled against its specific dependencies located in `/nix/store`.  Looking at what Nix installs, it has a wrapper script ensuring that utilities Bazel commonly uses are on `PATH`.  The actual executable (on my local install) has the following dependencies:
```
> ldd /nix/store/kqijqia5iiw6idnbkvxwbdw6y0jpdd30-bazel-6.4.0/bin/bazel-6.4.0-linux-x86_64  
        linux-vdso.so.1 (0x00007ffd1a711000)
        librt.so.1 => /nix/store/gqghjch4p1s69sv4mcjksb2kb65rwqjy-glibc-2.38-23/lib/librt.so.1 (0x00007f5f319d2000)
        libdl.so.2 => /nix/store/gqghjch4p1s69sv4mcjksb2kb65rwqjy-glibc-2.38-23/lib/libdl.so.2 (0x00007f5f319cd000)
        libpthread.so.0 => /nix/store/gqghjch4p1s69sv4mcjksb2kb65rwqjy-glibc-2.38-23/lib/libpthread.so.0 (0x00007f5f319c8000)
        libm.so.6 => /nix/store/gqghjch4p1s69sv4mcjksb2kb65rwqjy-glibc-2.38-23/lib/libm.so.6 (0x00007f5f318e8000)
        libstdc++.so.6 => /nix/store/9fy9zzhf613xp0c3jsjxbjq6yp8afrsv-gcc-12.3.0-lib/lib/libstdc++.so.6 (0x00007f5f31600000)
        libgcc_s.so.1 => /nix/store/9fy9zzhf613xp0c3jsjxbjq6yp8afrsv-gcc-12.3.0-lib/lib/libgcc_s.so.1 (0x00007f5f318c5000)
        libc.so.6 => /nix/store/gqghjch4p1s69sv4mcjksb2kb65rwqjy-glibc-2.38-23/lib/libc.so.6 (0x00007f5f31418000)
        /nix/store/gqghjch4p1s69sv4mcjksb2kb65rwqjy-glibc-2.38-23/lib/ld-linux-x86-64.so.2 => /nix/store/46m4xx889wlhsdj72j38fnlyyvvvvbyb-glibc-2.37-8/lib64/ld-linux-x86-64.so.2 (0x00007f5f319d9000)
```

I can make sure all of these utilities are available in the image and make sure all the dependencies are findable by `nix-ld`, but I'm concerned it will be brittle.

I think the more idiomatic thing to do would be mimic how `rules_nixpkgs` by invoking `nix shell` at the start of the test run to ensure `bazel` (and anything else we want) is present, and would be defined in `rules_rust`.  It looks like `bazelci.py` can support this, as it has a `setup` parameter and `shell_commands` that it looks like we can use to execute that ahead of the build.

**Comment by rickvanprim** [comment]
Ah I take that back, `bazelci.py` expects a `bazelisk` install, and expects it prior to running `setup`/`shell_commands`.  I will take another crack at getting a `bazelisk` container working 😅

**Comment by rickvanprim** [comment]
I've hit a small roadblock on getting this working with `nix-ld-rs` (which I just discovered is the upcoming replacement for `nix-ld` in Rust that better handles child processes).  The Nix version of Bazel has a few patches, so I'm not sure if making it past this current issue will still have other issues.

One (somewhat hacky) idea that comes to mind:  With `nix-ld-rs`, `bazelisk` is able to run, so I could make it to the part in `bazelci.py` where I could then install a Nix version of Bazel (defined in `rules_rust`, not here) which would shadow the `bazelisk` version, and then be the version used by the remainder of the CI run.  An immediately obvious downside would be if the `bazelisk` version advances to a version that's not available in Nix (like Bazel 7, presently).

Thoughts on what's acceptable/what you prefer @UebelAndre?

**Comment by illicitonion** [comment]
FWIW bazelisk can be pointed at a custom source of bazel binaries. So you could use bazelisk (which is statically linked Go so should work fine), and if you set `$BAZELISK_BASE_URL` or `$BAZELISK_FORMAT_URL` you could point to where Nix-built bazels are hosted.

**Comment by rickvanprim** [comment]
@meteorcloudy / @UebelAndre / @illicitonion, after a weekend of soul searching (and coming across https://github.com/OdysLam/ddapptools/blob/main/docker/Dockerfile), I have pivoted from a NixOS (esque) container to a standard Ubuntu container with Nix installed and dependencies fetched via Nix.  I've verified that the container is able to build the example without issue.

I think this is ready to merge, once any further feedback is resolved.  One question that comes to mind is, is `ubuntu-nix` a good name for this?  I could include the Ubuntu version but given that the packages are fetched from Nix, that distinction feels less relevant.  Also would an underscore be preferred to a dash?  The Docker tags use dashes but the `bazelci.py` names seem to use underscores (to differentiate platforms for the same image).

**Comment by meteorcloudy** [comment]
I think we probably need to also install google cloud sdk, otherwise it may fail when trying to upload test logs.

https://github.com/bazelbuild/continuous-integration/blob/85380e7b3c555dfa40911e87268cfed76e65e53f/buildkite/docker/ubuntu2204/Dockerfile#L64C30-L69

Do we need jdk, python3 as well?

**Comment by rickvanprim** [comment]
@meteorcloudy I've added `nixpkgs#google-cloud-sdk` and `nixpkgs#jdk17_headless` to the container.  Is there something I can run locally to verify?  Otherwise if you want to merge, I can follow up if an actual BuildKite run fails.

File: buildkite/bazelci.py [hunk 1]
```diff
@@ -482,11 +482,11 @@
         "docker-image": f"gcr.io/{DOCKER_REGISTRY_PREFIX}/fedora39-java17",
         "python": "python3",
     },
-    "nixos": {
-        "name": "NixOS",
-        "emoji-name": ":nixos:",
+    "ubuntu-nix": {
```
[the remark this answers is not in the document]
Comment (meteorcloudy) [reply buildkite/bazelci.py:485]: Ah, forgot to submit this comment.

**Comment by meteorcloudy** [comment]
@rickvanprim Nope, let's test in production ;)

File: buildkite/bazelci.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (rickvanprim) [reply buildkite/bazelci.py]: Done

**Review by meteorcloudy — APPROVED** [review]

**Comment by rickvanprim** [comment]
@meteorcloudy how long does it take to get built and pushed?  Also can you delete https://console.cloud.google.com/gcr/images/bazel-public/global/nixos now that it's renamed? :pray:

**Comment by meteorcloudy** [comment]
gcr.io/bazel-public/ubuntu2204-nix is pushed! and old images deleted

**Comment by rickvanprim** [comment]
Sigh, always something.  Evidently `~/.bashrc` isn't getting sourced.  I'll have a new PR up once I address that.

**Pull request merged**